app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationViewModel.kt

```diff
@@ -198,37 +209,73 @@ class ConversationViewModel @Inject constructor(
         }
     }
 
+    fun downloadAsset(messageId: String) {
+        viewModelScope.launch {
+            if (!isAssetDownloaded(messageId))
+                updateAssetMessageDownloadStatus(IN_PROGRESS, conversationId, messageId)
+
+            val result = getRawAssetData(conversationId, messageId)
+            updateAssetMessageDownloadStatus(if (result != null) DOWNLOADED else FAILED, conversationId, messageId)
+
+            if (result != null) {
+                val assetName = getAssetName(messageId)
+                showOnAssetDownloadedDialog(assetName, result)
+            }
+        }
+    }
+
+    fun showOnAssetDownloadedDialog(assetName: String?, assetData: ByteArray) {
+        conversationViewState = conversationViewState.copy(downloadedAssetDialogState = Displayed(assetName, assetData))
+    }
+
+    fun hideOnAssetDownloadedDialog() {
+        conversationViewState = conversationViewState.copy(downloadedAssetDialogState = Hidden)
+    }
+
+    private fun isAssetDownloaded(messageId: String): Boolean {
+        return conversationViewState.messages.firstOrNull {
+            it.messageHeader.messageId == messageId && it.messageContent is AssetMessage
+        }?.run {
+            (messageContent as AssetMessage).downloadStatus.run {
+                this == DOWNLOADED || this == IN_PROGRESS
+            }
+        } ?: false
+    }
+
     private fun getAssetLimitInBytes(): Int {
         // Users with a team attached have larger asset sending limits than default users
         return conversationViewState.userTeam?.run {
             ASSET_SIZE_TEAM_USER_LIMIT_BYTES
         } ?: ASSET_SIZE_DEFAULT_LIMIT_BYTES
     }
 
-
-    fun onError(errorCode: ConversationErrors) {
-        conversationViewState = conversationViewState.copy(onError = errorCode)
+    fun onSnackbarMessage(msgCode: ConversationSnackbarMessages) {
+        viewModelScope.launch {
+            // We need to reset the onSnackbarMessage state so that it doesn't show up again when going -> background -> resume back
+            conversationViewState = conversationViewState.copy(onSnackbarMessage = msgCode)
+            delay(SNACKBAR_MESSAGE_DELAY)
+            conversationViewState = conversationViewState.copy(onSnackbarMessage = null)
+        }
```

Comment:
i think a better place for doing this would be inside the UI, also why are we adding this magic delay ?

Reply:
I tried to keep as much logic away from the UI as possible. The delay is there to ensure the snackbar message will have enough time to be shown before it gets reset to null. I can also add that to the comments